src/libclipper/include/clipper/threadpool.hpp

```diff
@@ -0,0 +1,269 @@
+
+    /**
+     * Run the task.
+     */
+    virtual void execute() = 0;
+  };
+
+  template <typename Func>
+  class ThreadTask : public IThreadTask {
+   public:
+    ThreadTask(Func&& func) : func_{std::move(func)} {}
+
+    ~ThreadTask(void) override = default;
+    ThreadTask(const ThreadTask& rhs) = delete;
+    ThreadTask& operator=(const ThreadTask& rhs) = delete;
+    ThreadTask(ThreadTask&& other) = default;
+    ThreadTask& operator=(ThreadTask&& other) = default;
+
+    /**
+     * Run the task.
+     */
+    void execute() override { func_(); }
+
+   private:
+    Func func_;
+  };
+
+ public:
+  /**
+   * Constructor.
+   */
+  ThreadPool(void)
+      : ThreadPool{std::max(std::thread::hardware_concurrency(), 2u) - 1u} {
+    /*
+     * Always create at least one thread.  If hardware_concurrency() returns 0,
+     * subtracting one would turn it to UINT_MAX, so get the maximum of
+     * hardware_concurrency() and 2 before subtracting 1.
+     */
+  }
+
+  /**
+   * Constructor.
+   */
+  explicit ThreadPool(const std::uint32_t numThreads)
+      : done_{false}, work_queue_{}, threads_{} {
+    try {
+      for (std::uint32_t i = 0u; i < numThreads; ++i) {
+        threads_.emplace_back(&ThreadPool::worker, this);
+      }
+    } catch (...) {
+      destroy();
+      throw;
+    }
+  }
+
+  /**
+   * Non-copyable.
+   */
+  ThreadPool(const ThreadPool& rhs) = delete;
+
+  /**
+   * Non-assignable.
+   */
+  ThreadPool& operator=(const ThreadPool& rhs) = delete;
+
+  /**
+   * Destructor.
+   */
+  ~ThreadPool(void) { destroy(); }
+
+  /**
+   * Submit a job to be run by the thread pool.
+   */
+  template <typename Func, typename... Args>
+  auto submit(Func&& func, Args&&... args) {
+    auto boundTask =
+        std::bind(std::forward<Func>(func), std::forward<Args>(args)...);
+    using ResultType = std::result_of_t<decltype(boundTask)()>;
+    using PackagedTask = boost::packaged_task<ResultType()>;
+    using TaskType = ThreadTask<PackagedTask>;
+    PackagedTask task{std::move(boundTask)};
+    auto result_future = task.get_future();
+
+    work_queue_.push(std::make_unique<TaskType>(std::move(task)));
+    return result_future;
+  }
+
+ private:
+  /**
+   * Constantly running function each thread uses to acquire work items from the
+   * queue.
+   */
+  void worker(void) {
+    while (!done_) {
+      std::unique_ptr<IThreadTask> pTask{nullptr};
+      if (work_queue_.wait_pop(pTask)) {
+        pTask->execute();
+      }
+    }
+  }
+
+  /**
+   * Invalidates the queue and joins all running threads.
+   */
+  void destroy(void) {
+    done_ = true;
+    work_queue_.invalidate();
+    for (auto& thread : threads_) {
+      if (thread.joinable()) {
+        thread.join();
+      }
+    }
+  }
+
+ private:
+  std::atomic_bool done_;
+  ThreadSafeQueue<std::unique_ptr<IThreadTask>> work_queue_;
+  std::vector<std::thread> threads_;
+};
+
+namespace DefaultThreadPool {
```

Comment:
Consider using a name here that's more specific to the task executor. `TaskExecutionThreadPool`, for example. The description for the configuration argument specifying the number of threads in this pool is specific to task execution: *"Number of threads for the task execution threadpool"*.